src/analysis/calculus/inverse.lean

```diff
@@ -0,0 +1,582 @@
+  convert this,
+  ext x,
+  simp
+end
+
+end has_strict_fderiv_at
+
+namespace has_strict_fderiv_at
+
+variables [cs : complete_space (E × F)] {f : E × F → G} (f' : E × F →L[𝕜] G) (f'inv : G →L[𝕜] F)
+  {p : E × F} (hf : has_strict_fderiv_at f f' p)
+  (hf'l : ∀ y : F, f'inv (f' (0, y)) = y) (hf'r : ∀ z : G, f' (0, f'inv z) = z)
+
+/-- Formula for the derivative of an implicit function. -/
+def prod_implicit_function_fderiv : (E × G) →L[𝕜] F :=
+(f'inv.comp $ continuous_linear_map.snd 𝕜 E G -
+      f'.comp ((continuous_linear_map.id 𝕜 E).prod_map 0))
+
+variables {f' f'inv}
+
+@[simp] lemma prod_implicit_function_fderiv_apply (x) :
+  prod_implicit_function_fderiv f' f'inv x = f'inv (x.2 - f' (x.1, 0)) := rfl
+
+include hf'r
+
+lemma prod_implicit_fderiv_right_inverse (x : E) (z : G) :
+  f' (x, f'inv z) = f' (x, 0) + z :=
+by { conv_rhs { rw [← hf'r z] }, simp [← f'.map_add] }
+
+include hf'l
+
+variables (f' f'inv)
+
+/-- Derivative of an auxiliary function used in the proof of the implicit function theorem. -/
+def prod_implicit_function_aux_fderiv : (E × F) ≃L[𝕜] (E × G) :=
+continuous_linear_equiv.equiv_of_inverse
+  ((continuous_linear_map.fst 𝕜 E F).prod f')
+  ((continuous_linear_map.fst 𝕜 E G).prod $ prod_implicit_function_fderiv f' f'inv)
+  (λ ⟨x, y⟩, by simp [← continuous_linear_map.map_sub, hf'l])
+  (λ ⟨x, z⟩, by simp [-continuous_linear_map.map_sub, prod_implicit_fderiv_right_inverse hf'r])
+
+variables {f' f'inv}
+
+include hf
+
+lemma prod_implicit_function_aux_deriv :
+  has_strict_fderiv_at (λ x : E × F, (x.1, f x))
+    (prod_implicit_function_aux_fderiv f' f'inv hf'l hf'r : (E × F) →L[𝕜] E × G) p :=
+has_strict_fderiv_at_fst.prod hf
+
+include cs
+variable (f)
+
+/-- Implicit function `g` defined by an equation `f (x, g(x, y)) = z`. -/
+def prod_implicit_function (x : E × G) : F :=
+((hf.prod_implicit_function_aux_deriv hf'l hf'r).local_inverse _ _ _ x).2
+
+lemma prod_implicit_function_def (x : E × G) :
+  hf.prod_implicit_function f hf'l hf'r x =
+    ((hf.prod_implicit_function_aux_deriv hf'l hf'r).local_inverse _ _ _ x).2 :=
+rfl
+
+lemma to_prod_implicit_function :
+  has_strict_fderiv_at (hf.prod_implicit_function f hf'l hf'r)
+    (prod_implicit_function_fderiv f' f'inv) (p.1, f p) :=
+((hf.prod_implicit_function_aux_deriv hf'l hf'r).to_local_inverse).snd
+
+lemma eventually_apply_fst_prod_implicit_function_eq :
+  ∀ᶠ x in 𝓝 (p.1, f p), f ((x : E × G).1, hf.prod_implicit_function f hf'l hf'r x) = x.2 :=
+(hf.prod_implicit_function_aux_deriv hf'l hf'r).eventually_right_inverse.mono $
+  λ x hx, by { convert congr_arg prod.snd hx, convert prod.mk.eta,
+    exact (congr_arg prod.fst hx).symm }
+
+lemma eventually_prod_implicit_function_eq :
+  ∀ᶠ x in 𝓝 p, hf.prod_implicit_function f hf'l hf'r ((x : E × F).1, f x) = x.2 :=
+(hf.prod_implicit_function_aux_deriv hf'l hf'r).eventually_left_inverse.mono $
+  λ x hx, congr_arg prod.snd hx
+
+end has_strict_fderiv_at
+
```

Comment:
Another docstring with header here, explaining that you will reformulate the local inverse theorem for 1D functions? Also, maybe move before the implicit function theorem.